### ISSUE TITLE
First Attempt

### ISSUE THREAD

**Issue by davesmeghead** [msp1974/visonic_proxy#1]
Hi Mark
So I've had a quick go as follows:
I executed python run.py and this is the screen dump output
[screendump.zip](https://github.com/user-attachments/files/16383886/screendump.zip)

I tried connecting to it from my Home Assistant Integration in Standard Mode to see what would happen
I had to reload my Integration twice in HA as there seemed to be a connect problem but then it settled down.
I tried "Disarming" the panel from the VisonicGo App twice, both times it took a good couple of minutes for the command to get through, but it did eventually.

Why in here does it give different Panel Idents
```
2024-07-25 22:26:40.780 INFO     Disconnecting from Alarm P48621
2024-07-25 22:26:40.781 INFO     Client disconnected from Alarm P48621
2024-07-25 22:26:40.781 INFO     Disconnecting from Alarm_Monitor P49262
2024-07-25 22:26:40.781 INFO     Disconnecting from Alarm_Monitor P50188
2024-07-25 22:26:40.781 INFO     Client disconnected from Alarm_Monitor P49262
2024-07-25 22:26:40.781 INFO     Client disconnected from Alarm_Monitor P50188
2024-07-25 22:26:40.782 INFO     Shutting down connection to Visonic for P48621
```

I think you've got a double 0d going on, example
```
0d 0d a2 00 00 3f 00 00 00 00 00 00 00 43 da 0a e8 0a
```

Are you encoding the Non-B0 messages from Alarm_Monitor to the Panel as the Panel doesn't reply to them?

It's just highlighted to me that I need to decode more B0 data and also send more B0 data out.

**Comment by msp1974** [comment]
Ah yes, think a couple of issues here.

1. The message input was expecting from a2 to 43 and therefore adds checksum and wraps in 0d... 0a. I'll fix that as you are creating the whole message.
2. You are sending acks for each message you receive which in this setup is not rignt as they are already acked by the Visonic server.
3.  As you are in standard mode, it is also sending the wrong ack. You are sending 0d 02 fd 0a but it should be 0d 02 43 fd 0a (i know checksum will be different).

Do you want me to just filter out your ACKs or is it best for you to amend to not send them?

**Comment by davesmeghead** [comment]
1. OK
2. Agreed but see below
3. Agreed but see below 

Am I right in saying that in:
- Proxy mode (connected to Visonic Server) then the CM is better at filtering out the ACKs as they're not needed.  This does help us though as the CM knows that the message got through if the CM wanted to use that
- Standalone mode then the CM needs to use the ACKs to forward on (but might need to convert to Powerlink ACKs)

For a first pass though it did better than I expected so it is good progress.

**Comment by davesmeghead** [comment]
Just a note about checksums.
I've found that different panels calculate their checksums with a slightly different algorithm, older panels use _calculateCRC. 

For received messages I check against both, to send messages I use _calculateCRC as my PowerMax Panel ignores messages created using _calculateCRCAlt and it looks like all panels accept _calculateCRC 
See `MyChecksumCalc` in my pyhelper.py

```
    # alternative to calculate the checksum for sending and receiving messages
    def _calculateCRCAlt(self, msg: bytearray):
        """ Calculate CRC Checksum """
        # log.debug("[_calculateCRC] Calculating for: %s", self._toString(msg))
        # Calculate the checksum
        checksum = 0
        for char in msg[0 : len(msg)]:
            checksum += char
        # 29/8/2022
        #      This works for both my panels and always validates exactly (never using the +1 or -1 code in _validatePDU)
        #      It also matches the checksums that the Powerlink 3.1 module generates.
        checksum = 256 - (checksum % 255)
        if checksum == 256:
            checksum = 1
        # log.debug("[_calculateCRC] Calculating for: {self._toString(msg)}     calculated CRC is: {self._toString(bytearray([checksum]))}")
        return bytearray([checksum])

    # calculate the checksum for sending and receiving messages
    def _calculateCRC(self, msg: bytearray):
        """ Calculate CRC Checksum """
        # log.debug("[_calculateCRC] Calculating for: %s", self._toString(msg))
        # Calculate the checksum
        checksum = 0
        for char in msg[0 : len(msg)]:
            checksum += char
        checksum = 0xFF - (checksum % 0xFF)
        if checksum == 0xFF:
            checksum = 0x00
        # log.debug("[_calculateCRC] Calculating for: {self._toString(msg)}     calculated CRC is: {self._toString(bytearray([checksum]))}")
        return bytearray([checksum])
```

**Comment by msp1974** [comment]
Ok, so made the following changes

### General
1. Fixed the message wrapping issue as below
    ```  
    Messages can be injected in:
       Full format - 0d b0 01 6a 00 43 a0 0a \ 0d a2 00 00 08 00 00 00 00 00 00 00 43 12 0a
       Partial format - b0 01 6a 43 \ a2 00 00 08 00 00 00 00 00 00 00 43
       Short format (for B0 messages only) - b0 6a / b0 17 18 24
    ```
2. Updated the logging levels to make sense in the new segregated CM manager.  Leave on level 5 for now as we are testing.
    ```
    # Level of messages to be logged.
    # 1 just messages and errors/warnings
    # 2 same as 1 plus forward info
    # 3 same as 2 plus keepalives (if in non Proxy mode - in Proxy mode, same as 2)
    # 4 same as 3 plus acks
    # 5 same as 4 plus web messages
    # 6 same as 5 plus full PL31 raw message
    ```
### In Proxy Mode
1. ACKs received from the Alarm Monitor connection ie your code, will be dropped and not forwarded when in Proxy Mode.  As you say, we may use them when I look at some improved flow control.
2. Only requests made by your code will be ACKed back to you and these ACKs will not be sent to Visonic as it didnt request them.  
3. ACKs for Visonic requests will not be sent to you.  You will get just the Alarm repsonse.

### In Non Proxy Mode
1. The CM handles ACKs back to the Alarm and ignores your ACKs as in 2.  Do we think this is the right approach or should we forward ACKs from you?  It is just a config change in the Connection Profile definitions in visonic.py.  Make ```ack_received_messages``` False in ALARM connection profile and ```ignore_incomming_acks``` False in ALARM_MONITOR connection profile.  
2. Any ACKs you send are converted to the right format when received so they will send correctly.

----

> Just a note about checksums.

Yes I think that should be fine.  I am now (with last updates) only creating checksums for injected messages is not passed in full format.  So in the main, I am relying on the ones you calc.  I use the _calculateCRC method if needed.

----
Other issues I see

From the log output you sent me and some testing, ATM you need to make sure the CM is up and running and the Alarm has connected and sent its Hello messages before trying to send ones yourself.  I am going to look at a flow manager piece to hold these messages until it knows it has a connected and initiated alarm.  This will be part of the fix for the closed sessions - which is obvioulsy another issue.

It might also be worth trying to see how it functions with your code in non Proxy mode.

**Comment by davesmeghead** [comment]
For this 

> I am going to look at a flow manager piece to hold these messages until it knows it has a connected and initiated alarm

What about sending the Integration back a message that signifies "wait", the Integration could wait X seconds and try again. We could make this a generic wait that you could do at any time, and you could include X in the message.

For my first attempt, I didn't change any settings.  I left it like this for over 10 minutes and nothing else happened.  I did not attempt to connect the Integration.
```
2024-07-26 17:36:53.000 INFO     Starting Connection Manager
2024-07-26 17:36:53.000 INFO     Starting Webserver
2024-07-26 17:36:53.000 INFO     Started Watchdog Timer
2024-07-26 17:36:53.023 INFO     Listening for Alarm connection on 0.0.0.0 port 5001
2024-07-26 17:36:53.042 INFO     Listening for Alarm_Monitor connection on 0.0.0.0 port 5002
2024-07-26 17:36:53.058 INFO     Listening for Visonic_Monitor connection on 0.0.0.0 port 5003
2024-07-26 17:36:53.058 INFO     Connection Manager Running
2024-07-26 17:36:53.058 INFO     Message Coordinator Started
2024-07-26 17:37:15.393 INFO     Client connected to Alarm P49526 server. Clients: 1
2024-07-26 17:37:15.393 INFO     Initiating connection to Visonic on behalf of P49526
2024-07-26 17:37:15.431 INFO     Alarm 000000 P49526 MSGID:0046 -> 35 43 44 32 46 34 38 37 43 41 34 44 33 33 35 31 45 30 34 45 42 35 36 37 45 33 33 43 39 34 33 44 46 41 44 44 36 42 38 31 34 38 45 37 34 37 33 36 37 38 32 38 34 35 46 45 30 43 32 45 37 34 41 43 41 41 39 45 45 43 41 46 36 31 34 30 43 41 46 46 37 44 45 42 32 30 37 46 42 45 32 35 34 41 31 31 32 34 34 38 44 41 44 45 31 32 36 37 30 43 38 38 39 46 44 36 30 33 38 31 32 41 38 43 44 37 38
2024-07-26 17:37:15.431 INFO     Forwarding -> Visonic P49526
2024-07-26 17:37:15.443 WARNING  No connection to Visonic P49526 available to send message. Attempt 1
2024-07-26 17:37:15.500 INFO     Connected to Visonic server on port 5001 for P49526
2024-07-26 17:37:17.506 INFO     Visonic BD4100 P49526 MSGID:00R0 -> 5d 5f 31 36 3a 33 37 3a 31 37 2c 30 37 2d 32 36 2d 32 30 32
2024-07-26 17:37:17.506 INFO     Forwarding -> Alarm P49526
2024-07-26 17:37:17.506 WARNING  Unrecognised message format.  Skipping decoding
2024-07-26 17:37:17.507 INFO     Message: 0a 42 44 34 31 30 30 32 35 22 4e 41 4b 22 30 30 30 30 52 30 4c 30 41 30 5b 5d 5f 31 36 3a 33 37 3a 31 37 2c 30 37 2d 32 36 2d 32 30 32 34 0d
2024-07-26 17:37:17.507 INFO     Powerlink: PowerLink31Message(crc16='BD41', length='0025', type='NAK', msg_id='00R0', account_id='0A', panel_id='BD4100', message_class='5f', message=b']_16:37:17,07-26-202')
2024-07-26 17:37:17.543 INFO     Alarm 000000 P49526 MSGID:0046 -> 35 43 44 32 46 34 38 37 43 41 34 44 33 33 35 31 45 30 34 45 42 35 36 37 45 33 33 43 39 34 33 44 46 41 44 44 36 42 38 31 34 38 45 37 34 37 33 36 37 38 32 38 34 35 46 45 30 43 32 45 37 34 41 43 41 30 43 31 32 35 36 41 45 33 36 43 37 32 32 36 41 31 41 30 44 39 35 32 36 44 41 34 45 37 42 44 32 41 30 31 42 34 33 41 42 36 39 39 46 35 36 39 37 35 38 33 38 34 41 38 46 30 32 35 31 34 44
2024-07-26 17:37:17.544 INFO     Forwarding -> Visonic P49526
2024-07-26 17:37:17.611 INFO     Visonic 000000 P49526 MSGID:0046 -> 41 38 46 46 42 36 34 43 31 31 33 44 44 33 42 33 45 44 41 41 44 45 35 37 46 38 38 32 30 46 43 41 44 33 37 37 34 32 30 32 33 42 45 37 34 42 33 41 41 41 42 33 31 37 46 31 45 44 46 46 41 37 33
2024-07-26 17:37:17.611 INFO     Forwarding -> Alarm P49526
2024-07-26 17:37:17.616 INFO     Client disconnected from Alarm P49526
2024-07-26 17:37:17.616 INFO     Requesting Visonic P49526 to disconnect
2024-07-26 17:37:17.616 INFO     Shutting down connection to Visonic for P49526
2024-07-26 17:37:17.616 INFO     Visonic P49526 connection removed.  Remaining connections: [<ConnectionName.ALARM: 'Alarm'>, <ConnectionName.ALARM_MONITOR: 'Alarm_Monitor'>, <ConnectionName.VISONIC_MONITOR: 'Visonic_Monitor'>]
2024-07-26 17:37:17.617 INFO     Visonic P49526 server has disconnected
```

Remind me how to put it in Non-Proxy Mode?

**Comment by davesmeghead** [comment]
Found it
`PROXY_MODE = False`

**Comment by davesmeghead** [comment]
I've had a go at non-proxy mode and connected the Integration to it, I tried both my Standard and Powerlink Modes.  Both worked pretty well so we'll go with Powerlink Mode for the time being. 

[non-proxy-using-integration.zip](https://github.com/user-attachments/files/16394975/non-proxy-using-integration.zip)

Both actually created sensors in HA, Powerlink Mode was better / faster as it got the sensors above Zone 32 much quicker and it got the EPROM data.

By the looks of the output I need to start decoding the `b0 03 51` message and asking for them from the panel, that's my next task.  Time to work out how to do it.

**Comment by davesmeghead** [comment]
When in non-proxy how do I know when I can connect the integration to Alarm_Monitor?  Do I need to wait until it prints something?  If I don't I get this
```
2024-07-26 19:05:20.572 INFO     Starting Connection Manager
2024-07-26 19:05:20.573 INFO     Starting Webserver
2024-07-26 19:05:20.573 INFO     Started Watchdog Timer
2024-07-26 19:05:20.596 INFO     Listening for Alarm connection on 0.0.0.0 port 5001
2024-07-26 19:05:20.615 INFO     Listening for Alarm_Monitor connection on 0.0.0.0 port 5002
2024-07-26 19:05:20.616 INFO     Connection Manager Running
2024-07-26 19:05:20.616 INFO     Message Coordinator Started
2024-07-26 19:06:30.694 INFO     Client connected to Alarm_Monitor P45798 server. Clients: 1
2024-07-26 19:06:31.719 INFO     Alarm_Monitor None P45798 MSGID:0001 -> 0d 0b f4 0a
2024-07-26 19:06:31.721 INFO     Forwarding -> Alarm P45798
2024-07-26 19:06:31.721 WARNING  No connection to Alarm P45798 available to send message. Attempt 1
2024-07-26 19:06:33.439 INFO     Alarm_Monitor None P45798 MSGID:0002 -> 0d 0f f0 0a
2024-07-26 19:06:33.439 INFO     Forwarding -> Alarm P45798
2024-07-26 19:06:33.740 WARNING  No connection to Alarm P45798 available to send message. Attempt 2
2024-07-26 19:06:35.276 INFO     Alarm_Monitor None P45798 MSGID:0003 -> 0d 24 00 00 56 50 00 00 00 00 00 00 35 0a
2024-07-26 19:06:35.276 INFO     Forwarding -> Alarm P45798
2024-07-26 19:06:35.759 WARNING  No connection to Alarm P45798 available to send message. Attempt 3
2024-07-26 19:06:37.775 WARNING  No connection to Alarm P45798 available to send message. Attempt 4
2024-07-26 19:06:39.789 WARNING  No connection to Alarm P45798 available to send message. Attempt 5
2024-07-26 19:06:40.300 INFO     Alarm_Monitor None P45798 MSGID:0004 -> 0d 24 00 00 56 50 00 00 00 00 00 00 35 0a
2024-07-26 19:06:40.300 INFO     Forwarding -> Alarm P45798
2024-07-26 19:06:41.793 WARNING  No connection to Alarm P45798 available to send message. Attempt 6
2024-07-26 19:06:43.793 WARNING  No connection to Alarm P45798 available to send message. Attempt 7
2024-07-26 19:06:45.785 WARNING  No connection to Alarm P45798 available to send message. Attempt 8
2024-07-26 19:06:47.789 WARNING  No connection to Alarm P45798 available to send message. Attempt 9
2024-07-26 19:06:49.791 WARNING  No connection to Alarm P45798 available to send message. Attempt 10
2024-07-26 19:06:51.799 WARNING  No connection to Alarm P45798 available to send message. Attempt 11
2024-07-26 19:06:53.810 WARNING  No connection to Alarm P45798 available to send message. Attempt 12
2024-07-26 19:06:55.827 WARNING  No connection to Alarm P45798 available to send message. Attempt 13
2024-07-26 19:06:57.848 WARNING  No connection to Alarm P45798 available to send message. Attempt 14
2024-07-26 19:06:59.854 WARNING  No connection to Alarm P45798 available to send message. Attempt 15
```

Is there something I can to to get it connected or do I just wait?

**Comment by msp1974** [comment]
The messages it is sending is some sort of Alarm message.  In the Plink message wrapper it is *ADM-CID and wants a *ACK back.  Everything else is VIS-BBA or VIS-ACK with our normal message structure.  It is expecting a response which we are not sending, so it is disconnecting.  I'm not sure how to decode this mesage or create this response at the moment. If you put it in Proxy mode until it stops sending this then it will work.  I'll have a look to see if mine sends the same message and therefore I can just get the response from PM server and inject that back in non Proxy mode.

**Comment by davesmeghead** [comment]
I've got it in proxy mode and most of the time the proxy doesn't connect, it just sits there like this
```
2024-07-26 22:27:31.586 INFO     Starting Connection Manager
2024-07-26 22:27:31.586 INFO     Starting Webserver
2024-07-26 22:27:31.586 INFO     Started Watchdog Timer
2024-07-26 22:27:31.607 INFO     Listening for Alarm connection on 0.0.0.0 port 5001
2024-07-26 22:27:31.627 INFO     Listening for Alarm_Monitor connection on 0.0.0.0 port 5002
2024-07-26 22:27:31.646 INFO     Listening for Visonic_Monitor connection on 0.0.0.0 port 5003
2024-07-26 22:27:31.646 INFO     Connection Manager Running
2024-07-26 22:27:31.646 INFO     Message Coordinator Started
```

On the few occasions I can get it to connect my poor little Integration just gets flooded with B0 data, this is an example:
[over 2000 b0 in 8 minutes.zip](https://github.com/user-attachments/files/16396839/over.2000.b0.in.8.minutes.zip)

`LOG_TO_FILE = False` is false by default so I don't have your log file but text was scrolling up the screen pretty quickly.

I've only got it to connect a few (4 or 5) times and a couple of those times I unblocked the panels ip in the asus router, saw it working and blocked it again quickly.

**Comment by msp1974** [comment]
That seems to me like maybe an issue with the web service.  It basically connects to that and requires a response before it will connect on port 5001.  Ok, few questions:

1. What logging level do you have it set on in const.py?   None of your logs show the https logs which either means logging level is below 5 or there isnt any.  Can you make sure 5 and see if you see https logs? And share if so.
2. If above is a no, that is the issue.  I did see that this used to be non ssl on port 8081 (i think).  Can you unblock your panel on your router and see what connections it is making.  I can see this in system logs -> connections on  my asus router.
3. Can you confirm your powerlink sw version.

**Comment by msp1974** [comment]
Just to add, both of my panels connect reliably within 10s or so, so something is different here.

**Comment by msp1974** [comment]
AHHH, think I just worked it out.  I had to generate some certificates for the SSL webserver and github complained about me having them in the repo, so I took them out and have in the dockerfile to create them on build.  But you are running in non docker, so they will be missing.  I will add a script to create them and put in the right place.  You can then run this.

**Comment by davesmeghead** [comment]
I've been using proxy and it works OK apart from 3 things:
- After a while it receives the AD command you spoke about and just disconnects
- After a (different while) the VisonicGo app sometimes goes back to the Panels List and my Panel is no longer there, I have to add it again, and them 50/50 on whether it reconnects
- Using the Integrations Powerlink Mode just doesn't work with proxy so I connect with Standard Mode.  2 reasons:
   - The Integration expects to "manage" the powerlink connection and it just fails e.g. Enrol doesn't happen etc etc
   - The Integration expect to download the EPROM, as soon as it puts the panel in to Downloading the Visonic Server detects this and pushes it back out, so the EPROM download fails

I'm running it all on my Windows PC at the moment so I created the certificates manually and put them in place.

The 4th thing still remains, decode more B0 messages :)

**Comment by msp1974** [comment]
Yes, currently working on how I manage these differently to alway keep the alarm connection but connect and disconnect the Visonic as needed.  You will find if you are in the app, this keeps it alive also.  Just put that in the instructions for the integration - it'll be fine! :-)

> After a (different while) the VisonicGo app sometimes goes back to the Panels List and my Panel is no longer there, I have to add it again, and them 50/50 on whether it reconnects. EDIT: I can still use it from my Integration though.

I've not seen that - very wierd as that is what you have registered on your account with Visonic.  Is this your main or test alarm?  Is it possibly registered with someone else also and they remove it?  Have you tried it in the installer app - does it do the same?

**Comment by msp1974** [comment]
Closing as all issues described here are resolved.
